pkgs/http/lib/src/browser_client.dart

```diff
@@ -155,50 +155,105 @@ Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
     }
     e = ClientException(message, request.url);
   }
-  Error.throwWithStackTrace(e, st);
+  return e;
+}
+
+Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
+  Error.throwWithStackTrace(_toClientException(e, request), st);
 }
 
-Stream<List<int>> _readBody(BaseRequest request, Response response) async* {
+Stream<List<int>> _readBody(BaseRequest request, Response response) {
   final bodyStreamReader =
       response.body?.getReader() as ReadableStreamDefaultReader?;
-
   if (bodyStreamReader == null) {
-    return;
+    return const Stream.empty();
   }
 
-  var isDone = false, isError = false;
-  try {
-    while (true) {
-      final chunk = await bodyStreamReader.read().toDart;
+  final controller = StreamController<List<int>>();
+  final cancelCompleter = Completer<Null>();
+  Completer<void>? waitingForResume;
+  var readerEmittedDone = false;
+
+  Future<void> readUntilDoneOrCancelled() async {
+    while (!cancelCompleter.isCompleted) {
+      assert(waitingForResume == null);
+
+      final chunk = await Future.any(
+          [bodyStreamReader.read().toDart, cancelCompleter.future]);
+      if (chunk == null) {
+        // Stream subscription was cancelled. We'll cancel the reader later.
+        assert(cancelCompleter.isCompleted);
+        return;
+      }
+
       if (chunk.done) {
-        isDone = true;
-        break;
+        readerEmittedDone = true;
+        return;
       }
-      yield (chunk.value! as JSUint8Array).toDart;
-    }
-  } catch (e, st) {
-    isError = true;
-    _rethrowAsClientException(e, st, request);
-  } finally {
-    if (!isDone) {
-      try {
-        // catchError here is a temporary workaround for
-        // http://dartbug.com/57046: an exception from cancel() will
-        // clobber an exception which is currently in flight.
-        await bodyStreamReader
-            .cancel()
-            .toDart
-            .catchError((_) => null, test: (_) => isError);
-      } catch (e, st) {
-        // If we have already encountered an error swallow the
-        // error from cancel and simply let the original error to be
-        // rethrown.
-        if (!isError) {
-          _rethrowAsClientException(e, st, request);
-        }
+
+      controller.add((chunk.value! as JSUint8Array).toDart);
+      if (controller.isPaused) {
+        final resume = waitingForResume = Completer();
```

Comment:
(Good choice to only allocate completer if actually needed, and not for each pause/resume!)